crates/iota-genesis-builder/src/stake.rs

```diff
@@ -105,168 +103,304 @@ impl GenesisStake {
+            if timelocks_pool.is_none() && gas_coins_pool.is_none() {
+                anyhow::bail!("no timelocks or gas-coin objects found for delegator {delegator:?}");
+            }
+            stake.delegate_genesis_stake(
+                &validators_allocations,
+                delegator,
+                &mut timelocks_pool.unwrap_or_default().into_iter(),
+                &mut gas_coins_pool
+                    .unwrap_or_default()
+                    .into_iter()
+                    .map(|object| (object, 0)),
+            )?;
+        }
+
+        Ok(stake)
+    }
+
+    fn create_token_allocation(
+        &mut self,
+        recipient_address: IotaAddress,
+        amount_nanos: u64,
+        staked_with_validator: Option<IotaAddress>,
+        staked_with_timelock_expiration: Option<u64>,
+    ) {
+        self.token_allocation.push(TokenAllocation {
+            recipient_address,
+            amount_nanos,
+            staked_with_validator,
+            staked_with_timelock_expiration,
+        });
+    }
+
+    /// Create the necessary allocations for `validators_allocations` using the
+    /// assets of the `delegator`.
+    ///
+    /// This function iterates in turn over [`TimeLock`] and
+    /// [`GasCoin`][iota_types::gas_coin::GasCoin] objects created
+    /// during stardust migration that are owned by the `delegator`.
+    pub fn delegate_genesis_stake<'obj>(
+        &mut self,
+        validators_allocations: &[ValidatorAllocation],
+        delegator: IotaAddress,
+        timelocks_pool: &mut impl Iterator<Item = (&'obj Object, ExpirationTimestamp)>,
+        gas_coins_pool: &mut impl Iterator<Item = (&'obj Object, ExpirationTimestamp)>,
+    ) -> anyhow::Result<()> {
+        // Temp stores for holding the surplus
+        let mut timelock_surplus = CoinSurplus::default();
+        let mut gas_surplus = CoinSurplus::default();
+
+        // Then, try to create new token allocations for each validator using the
+        // objects fetched above
+        for validator_allocation in validators_allocations {
+            // The validaotr address
+            let validator = validator_allocation.validator;
+            // The target amount of nanos to be staked, either with timelock or gas objects
+            let mut target_stake = validator_allocation.amount_nanos_to_stake;
+            // The gas to pay to the validator
+            let gas_to_pay = validator_allocation.amount_nanos_to_pay_gas;
+
+            // Start filling allocations with timelocks
+
+            // Pick fresh timelock objects (if present) and possibly reuse the surplus
+            // coming from the previous iteration
+            let mut timelock_objects =
+                pick_objects_for_allocation(timelocks_pool, target_stake, &mut timelock_surplus);
+            if !timelock_objects.to_burn.is_empty() {
+                // Inside this block, the previous surplus is empty so we update it (possibly)
```

Comment:
We are technically "draining" the surplus object from its surplus_nanos. Therefore, the surplus object can be burned. 

In the `pick_objects_for_allocation` function, we try and take all the nanos that we need from the surplus object first. If we have "drained" the surplus object, we place it in the `timelock_objects.to_burn` Vec. This means that the previous surplus is empty, and it needs to be replaced with a new one (picked in `pick_objects_for_allocation` because the previous allocation might have needed some extra nanos). 

Inside pick_objects_for_allocation this happens:
surplus = Coin(50)
1. if target_stake == 49 then  surplus = Coin(1); timelock_objects.amount_nanos = 49; [timelock_objects.to](http://timelock_objects.to/)_burn = empty
2. if target_stake == 50 then  surplus = empty; timelock_objects.amount_nanos = 50; [timelock_objects.to](http://timelock_objects.to/)_burn = [surplus]
3. if target_stake == 51 then  surplus = empty; timelock_objects.amount_nanos = 51; [timelock_objects.to](http://timelock_objects.to/)_burn = [surplus, timelock_1]

> Btw, shouldn't we check that it is really empty before replacing it?

This is a given, since we're in the scope of `timelock_objects.to_burn.is_empty()`. If the `to_burn` field is empty, it means we still have nanos in the surplus object. Would you advice adding a check here out of safety/clarity? Or could some comments explaining this be sufficient?

> Another thing that comes to my mind.... why do we even "burn" something? Is it because we consume those "UTXO's"? If yes, shouldn't we rather call it to_consume instead of to_burn?
Burning, at least in my head, is only used if we actually destroy tokens that are gone forever afterwards.

Indeed, at first glance I also think we could use another term for this, something more in line with what's being used throughout the code: to `destroy` empty objects. 
However, these objects are not really "consumed" in the actual object database yet. This happens in a later stage in the `build_unsigned_genesis_data` phase of the ceremony. First, the required genesis objects are created. After this, the genesis objects are being "cleansed" from the to be burned objects that we have specified earlier.

Reply:
Changed burn into destroy here [1ef83d2](https://github.com/iotaledger/iota/pull/4346/commits/1ef83d2ed49087ab6f4e545de2316de1cf59229b)